src/main/scala/sigmastate/serialization/ErgoTreeSerializer.scala

```diff
@@ -18,6 +21,23 @@ object ErgoTreeSerializer {
     w.toBytes
   }
 
+  def serialize(tree: Value[SType]): Array[Byte] = {
+    val constantStore = new ConstantStore()
+    val treeWriter = Serializer.startWriter(constantStore)
+    ValueSerializer.serialize(tree, treeWriter)
+    val extractedConstants = constantStore.getAll
```

Comment:
@greenhat looks like we are doing serialization twice?

Reply:
@ergomorphic Not exactly. We first extract constants and serialize the tree (with placeholders). Than, serialize just the extracted constants.